src/numerics/petsc_matrix.C

```diff
@@ -1201,6 +1201,53 @@ void PetscMatrix<T>::add (const T a_in, const SparseMatrix<T> & X_in)
 }
 
 
+template <typename T>
+void PetscMatrix<T>::mat_mult (SparseMatrix<T> & X_in, SparseMatrix<T> & Y_out)
+{
+  libmesh_assert (this->initialized());
+
+   // sanity check.
+   libmesh_assert_equal_to (this->n(), X_in.m());
+   libmesh_assert_equal_to (this->m(), Y_out.m());
+   libmesh_assert_equal_to (X_in.n(), Y_out.n());
+
+   const PetscMatrix<T> * X = cast_ptr<const PetscMatrix<T> *> (&X_in);
+   PetscMatrix<T> * Y = cast_ptr<PetscMatrix<T> *> (&Y_out);
+
+   libmesh_assert (X);
+   libmesh_assert (Y);
+
+   PetscErrorCode ierr=0;
+
+   // the matrix from which we copy the values has to be assembled/closed
+   libmesh_assert(X->closed());
+
+   semiparallel_only();
+
+   ierr = MatMatMult(_mat, X->_mat, MAT_INITIAL_MATRIX, PETSC_DEFAULT, &Y->_mat);
```

Comment:
> We should leave `MatReuse` as an in parameter, and set its default value as `MAT_INITIAL_MATRIX`
> 
> For some cases, if your matrix sparsity does not change, it is a big deal to reuse the sparsity pattern using `MAT_REUSE_MATRIX`

I am keeping a virtual version in `sparse_matrix.h`:

```
virtual void matrix_matrix_mult (SparseMatrix<T> & /*X*/, SparseMatrix<T> & /*Y*/, const MatReuse & /*Reuse=MAT_INITIAL_MATRIX*/)
  { libmesh_not_implemented(); }
```

Therefore, I included the `petscmat.h` in the header. It compiles fine on my own machine while causing some failed builds on GitHub (complaining about not being able to find the `petscmat.h`). Any ideas on how to eliminate the error here? @fdkong

Reply:
We can't introduce the PETSc type `MatReuse` at the base class level (and we can't include PETSc headers without properly guarding them) since libMesh supports `--disable-petsc` builds. You can work around this issue by creating a new enum in libmesh which you can then map to the PETSc enumeration values. (see, for example: enum_convergence_flags.h). Since this is kind of an optional/nice-to-have feature, I'd say you could skip it in your current PR and then myself or one of the other devs can go back and implement it a bit later, once your current PR is merged.

Reply:
@dewenyushu Do you like to address John's comments? Just introduce `enum`.

It is will be very useful because you reuse expensive calculations during the preconditioner application.